src/pages/patientView/timeline2/TimelineWrapperUtils.ts

```diff
@@ -0,0 +1,80 @@
+import { TimelineEvent } from 'cbioportal-clinical-timeline';
+import { ISampleMetaDeta } from 'pages/patientView/timeline2/TimelineWrapper';
+import _ from 'lodash';
+
+const DEFAULT_LABEL = '-';
+
+export function getSampleInfo(
+    event: TimelineEvent,
+    caseMetaData: ISampleMetaDeta
+) {
+    const sampleId = event.event.attributes.find(
+        (att: any) => att.key === 'SAMPLE_ID'
+    );
+    if (sampleId) {
+        const color = caseMetaData.color[sampleId.value] || '#333333';
+        const label = caseMetaData.label[sampleId.value] || DEFAULT_LABEL;
+
+        return { color, label };
+    }
+
+    return null;
+}
+
+export function getNumberRangeLabel(sortedNumbers: number[]) {
+    if (!sortedNumbers.length) {
+        return DEFAULT_LABEL;
+    }
+
+    // aggregate into ranges
+    const ranges: { start: number; end: number }[] = [];
+    let currentRange = {
+        start: sortedNumbers[0],
+        end: sortedNumbers[0],
+    };
+    for (let i = 1; i < sortedNumbers.length; i++) {
```

Comment:
i think we should establish a rule of never using native for loops unless there's an optimization case.   i find them harder to read because you have to use the index to get element and also, there's classic issue where length is read every time unecessarily. granted here it's negligible.

Reply:
I agree with you but in this case I need to start with the second element. would be sweet if there was a nicer way to do that besides if index ===0 continue